core/controllers/load_functions.py

```diff
@@ -0,0 +1,257 @@
+
+    The function first retrieves the relevant data from the workbook using the provided mapping.
+    It then processes the data to get outcomes and outputs to insert,
+    and finally adds all the models to the database session.
+
+    :param workbook: a dictionary of DataFrames of table data to be inserted into the db.
+    :param mapping: the mapping of the relevant DataFrame to its attributes as they appear in the db.
+    """
+    worksheet = workbook[mapping.table]
+    models = mapping.map_data_to_models(worksheet)
+
+    models = get_outcomes_outputs_to_insert(mapping, models)
+    db.session.add_all(models)
+
+
+def generic_load(workbook: dict[str, pd.DataFrame], mapping: DataMapping, submission_id):
+    """
+    Generic function to load data into a specified table.
+
+    The function takes a workbook, a mapping of relevant data, and a submission ID.
+    It adds the submission ID to the worksheet, maps the data to models using the provided mapping,
+    and adds all the models to the database session.
+
+    :param workbook: a dictionary of DataFrames of table data to be inserted into the db.
+    :param mapping: the mapping of the relevant DataFrame to its attributes as they appear in the db.
+    :param submission_id: the ID of the submission associated with the data.
+    """
+    worksheet = workbook[mapping.table]
+    worksheet["Submission ID"] = submission_id
+    models = mapping.map_data_to_models(worksheet)
+
+    db.session.add_all(models)
+
+
+def get_submission_id(workbook: dict[str, pd.DataFrame]) -> str:
+    """
+    Retrieves or generates a submission ID based on the information in the provided workbook.
+
+    The function extracts the reporting round and programme ID from the workbook.
+    If a matching programme exists in the same reporting round, it finds the corresponding project and submission.
+    The submission ID is then retrieved, and the existing submission is removed.
+    If there is no matching programme for the current round, a new submission ID is generated.
+
+    :param workbook: a dictionary of DataFrames containing table data to determine the submission ID.
+    :return: a string representing the submission ID.
+    """
+    reporting_round = int(workbook["Submission_Ref"]["Reporting Round"].iloc[0])
+    programme_id = workbook["Programme_Ref"]["Programme ID"].iloc[0]
+    # if already added this round, this entity used to drop existing round data
+    programme_exists_same_round = (
+        Programme.query.join(Project)
+        .join(Submission)
+        .filter(Programme.programme_id == programme_id)
+        .filter(Submission.reporting_round == reporting_round)
+        .first()
+    )
+
+    if programme_exists_same_round:
+        matching_project = None
+        for project in programme_exists_same_round.projects:
+            if project.submission.reporting_round == reporting_round:
+                matching_project = project
+                break
+        if matching_project:
```

Comment:
this logic definitely needs splitting out further. Deleting a submission shouldn't come under the scope of "get_submission_id"

Reply:
So somewhere like, in higher level function (populate_db) check if submission id exists in a function and return that id or None if not.
Then, also from high level function, if exists already (id not None) call another function to do the delete
Then, call next_submission_id if None
Then continue with insert as normal.

Does that make sense? HAving the workflow slpit up more allows all the components to be more easily unit tested / mocked, and also makes it clearer what the scope of each bit does based on the function names etc

Reply:
Yep, makes sense. Can split that out.

Reply:
I've abstracted out the deletion based on submission to a separate function. However, the code used to assess whether to retrieve a submission id if it exists is the same as the code used to determine if we should delete all of a submission's children. I've now got a new 'master' function - upsert_or_get_submission_id() - which calls the deletion function if it needs to upsert. This way we can test the functionality separately without repeating code. 

The other possibility is repeating the code that does the checks if the submission id already exists for a given round.